### PR TITLE
feat: Added new language code standard

### PR DESCRIPTION
**Pull request by KennethEnevoldsen**
This adds a new language code standard using an iso 639-3 code as well as an iso code for scripts. 

I believe the benchmarks are currently preserved as they should be. I also believe all tasks run as intended (I have tested a few of the relevant candidates).

A few things we would probably need to fix:
- [ ] MTEB(task_langs=[...]), currently uses just the user-specified splits, we should instead transfer it to use the standard language codes.
  - Currently the above filter also only filters within tasks (not monolingual tasks). I believe it should do both.
  - While we are changing it, we could also add a few other filters based on domains etc.
- [ ] The above will cause a need to update the benchmark lists (shouldn't be too big)
- [ ] This change does not lead to changes (at least as I understand it) in the results, but it probably should. We have already discussed standardizing the format of the results.

The suggestions above will be major breaking which changes the way the benchmark interface works as well as how results are stored. 

supersedes: #319
fixes: #251

### CODE REVIEW

File: mteb/abstasks/languages.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import json
+from pathlib import Path
+
+path_to_lang_codes = Path(__file__).parent / "iso_639_3_to_language.json"
+path_to_lang_scripts = Path(__file__).parent / "iso_15924_to_script.json"
+
+
+with path_to_lang_codes.open("r") as f:
+    ISOCODE639_3_TO_LANGUAGE = json.load(f)
```
Comment (MartinBernstorff) [inline mteb/abstasks/languages.py:14]: I'd shorten this to ISO_TO_LANGUAGE.

File: mteb/abstasks/languages.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+with path_to_lang_scripts.open("r") as f:
+    ISOCODE15924_TO_SCRIPT = json.load(f)
```
Comment (MartinBernstorff) [inline mteb/abstasks/languages.py:17]: Same as above.

File: mteb/abstasks/AbsTask.py [hunk 3]
```diff
@@ -58,3 +58,24 @@ def evaluate(self, model, split="test"):
         :param split: Which datasplit to be used.
         """
         raise NotImplementedError
+
+    @property
+    def languages(self) -> set[str]:
```
Comment (MartinBernstorff) [inline mteb/abstasks/AbsTask.py:63]: My impression is that providing these "shortcuts" tend to be annoying for maintenance, but your mileage may vary.

File: mteb/abstasks/TaskMetadata.py [hunk 4]
```diff
@@ -92,7 +95,8 @@
 ]  # Allows the type to be a string, but ensures that the string is a valid date
 
 SPLIT_NAME = str
-
+LANGUAGE_SCRIPT_CODE = str  # a 3-letter ISO 639-3 language code followed by a 4-letter ISO 15924 script code (e.g. "eng-Latn")
```
Comment (MartinBernstorff) [inline mteb/abstasks/TaskMetadata.py:98]: ISO_LANGUAGE_SCRIPT? Follows the naming of the other type aliases.

File: mteb/abstasks/TaskMetadata.py [hunk 5]
```diff
@@ -180,3 +187,58 @@ def _check_dataset_revision_is_specified(cls, dataset):
                 "It is encourage to specify a dataset revision for reproducability"
             )
         return dataset
+
+    @field_validator("eval_langs")
+    def _check_eval_langs(cls, eval_langs):
+        """
+        This method checks that the eval_langs are specified as a list of languages.
+        """
+        if isinstance(eval_langs, list):
```
Comment (MartinBernstorff) [inline mteb/abstasks/TaskMetadata.py:196]: If someone specifies a tuple, we are not covering. Perhaps have the first check a dict, the rest check sequences?

File: mteb/abstasks/TaskMetadata.py [hunk 6]
```diff
@@ -180,3 +187,58 @@ def _check_dataset_revision_is_specified(cls, dataset):
...
+    @staticmethod
+    def check_language_code(code):
+        """
+        This method checks that the language code (e.g. "eng-Latn") is valid.
+        """
+        lang, script = code.split("-")
+        if lang not in ISOCODE639_3_TO_LANGUAGE:
+            raise ValueError(f"Invalid language code: {lang}")
+        if script not in ISOCODE15924_TO_SCRIPT:
+            raise ValueError(f"Invalid script code: {script}")
```
Comment (MartinBernstorff) [inline mteb/abstasks/TaskMetadata.py:214]: Excellent check! Might be worth it to point the user towards the valid values? E.g. point them to the file containing all valid scripts/langs.

File: mteb/abstasks/TaskMetadata.py [hunk 7]
```diff
@@ -180,3 +187,58 @@ def _check_dataset_revision_is_specified(cls, dataset):
...
+            for code in eval_langs:
+                cls.check_language_code(code)
+        elif isinstance(eval_langs, dict):
+            for langs in eval_langs.values():
+                for code in langs:
+                    cls.check_language_code(code)
+        return eval_langs
+
+    @staticmethod
+    def check_language_code(code):
```
Comment (MartinBernstorff) [inline mteb/abstasks/TaskMetadata.py:206]: Private?

File: mteb/evaluation/MTEB.py [hunk 8]
```diff
@@ -41,20 +41,14 @@ def __init__(
...
-        task_langs: list of str / None
-            List of languages to be evaluated. if None, all languages will be evaluated. ["en", "de"] will evaluate on "en", "de", "en-de" and "de-en"
-        tasks: list of str / None
-            List of tasks to be evaluated. If specified, we filter tasks based on `task_langs` only
-        version: int / None
-            Version of the benchmark to use. If None, latest is used
-        err_logs_path: str / None
-            Path to save error logs
+        Args:
+            task_types: List of task types (Clustering, Retrieval..) to be evaluated. If None, all tasks will be evaluated
```
Comment (MartinBernstorff) [inline mteb/evaluation/MTEB.py:45]: Since this takes a list of [str], we might want to specify where to find valid inputs.

File: mteb/evaluation/MTEB.py [hunk 9]
```diff
@@ -41,20 +41,14 @@ def __init__(
...
+            task_categories: List of task categories (s2s, p2p..) to be evaluated. If None, all tasks will be evaluated
```
Comment (MartinBernstorff) [inline mteb/evaluation/MTEB.py:46]: Same as above.

File: mteb/evaluation/MTEB.py (hunk 8 above)
Comment (MartinBernstorff) [reply mteb/evaluation/MTEB.py:45]: Ah, I see you only did formatting changes. Feel free to disregard if this is acceptable.

**Review by MartinBernstorff — APPROVED** [review]
Good work @Kenneth! Huge diff though, so might very well have missed something, but the stuff I noticed all seems very minor.

File: mteb/abstasks/AbsTask.py (hunk 3 above)
Comment (KennethEnevoldsen) [reply mteb/abstasks/AbsTask.py:63]: Annoying for maintenance and nice for use. My guess would be that it removes confusion due to the mapping dicts (leading to less issues). All a guess though. I will need to for:

```
MTEB(task_langs=[...]), currently uses just the user-specified splits, we should instead transfer it to use the standard language codes.
```
So rather than implement it in the MTEB object I believe it is better places here.

**Comment by KennethEnevoldsen** [comment]
I will merge this in now due to the many merge conflicts I keep having to solve. Once it is in we can basically launch MMTEB as the other changes are only nice to have, but not need to have and don't cause issues for people adding datasets

**Comment by Muennighoff** [comment]
Do we have to change the language codes in all existing result files?

**Comment by KennethEnevoldsen** [comment]
@Muennighoff atm. I don't believe it changes the results file at all (still uses the hf_name for the multilingual dataset e.g. "fr-en") and for the monolingual datasets, it doesn't include a lang-id. The same is the case for MTEB(task_langs="fr-en"), which should use the same mappings as before and still match the hf-name.
